stm32: mcuboot: add support for mcuboot tests and samples

djiatsaf-st opened the pull request:
This PR adds support for testing MCUboot tests and samples by updating the MCUboot Kconfig defaults. STM32 boards will no longer use `SWAP_USING_MOVE` as the default swap mode; instead, `SWAP_USING_OFFSET` becomes the default for STM32.

Following the remarks and suggestions made in PR #93862, all changes aim to comply with the rules outlined in that review (https://github.com/zephyrproject-rtos/zephyr/pull/93862#pullrequestreview-3320196653).

Related tests and samples:

- zephyr/samples/sysbuild/with_mcuboot
- zephyr/tests/boot/test_mcuboot
- zephyr/tests/boot/with_mcumgr


Fixes : https://github.com/zephyrproject-rtos/zephyr/issues/93329

djiatsaf-st committed:
modules: kconfig: mcuboot: make SWAP_USING_OFFSET the default for STM32

Update MCUboot configuration so that STM32 no longer defaults to
SWAP_USING_MOVE. Instead, SWAP_USING_OFFSET becomes the default mode
for STM32 boards.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
boards: st: add support for boards compatible with swap using offset

- Swap using offset requires that slot1 has one extra sector size
compared to slot0

- Additionally, add a storage partition with at least 3 sectors for
NVS if needed.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
boards: st: add support for boards compatible with swap using scratch

SomeSTM32 series, like F2, F4, and F7, do not have homogeneous flash
sector sizes; they include very large sectors such as 128 KB
and even 256 KB.

The scratch partition needs to be large enough to hold the largest
sector size ( 256k for F7 and 128k for F2).

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
samples: sysbuild: with_mcuboot: update sample.yaml file

Add STM32 supported boards for the test.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
tests: boot: test_mcuboot: update testcase.yaml file

Add STM32 supported boards for the test.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
tests: boot: with_mcumgr: update testcase.yaml file

Add STM32 supported boards for the test.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

etienne-lms commented on boards/st/nucleo_h753zi/nucleo_h753zi.dts:
```diff
@@ -261,13 +261,13 @@ zephyr_udc0: &usbotg_fs {
+		storage_partition: partition@c0000 {
 			label = "storage";
-			reg = <0x000a0000 DT_SIZE_K(256)>;
+			reg = <0x000c0000 DT_SIZE_K(256)>;
```
Why not using 3 sectors even if it spreads into bank2 1st sector?
Even, why not using the overall 2MB flash, increasing slot0/slot1 sizes by several 128kB sectors?

djiatsaf-st committed:
tests: drivers: flash: stm32: handle conflict for storage partition

Since the addition of a new storage partition definition in
the nucleo_f746zg dts, there is no need to have this configuration
in the overlay file.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

djiatsaf-st committed:
release: migration-guide: 4.4: add entry about mcuboot mode

STM32 platforms now use the default MCUboot operating mode
swap using offset instead of swap using move.

Signed-off-by: Fabrice DJIATSA <fabrice.djiatsa-ext@st.com>

JarmouniA commented on boards/st/nucleo_c071rb/nucleo_c071rb.dts:
```diff
@@ -194,3 +195,32 @@ zephyr_udc0: &usb {
+			reg = <0x00013800 DT_SIZE_K(44)>;
+		};
+
+		/* Set 6KB of storage at the end of 128KB flash */
```
```suggestion
```
Nit: not really useful.
Same for occurrences in other files.

asm5878 reviewed (commented):
Hello @djiatsaf-st , could you please be so kind to add a flash partioning for WBA6 (nucleo and DK) ?

I could make a dedicated PR, but it will be better making these changes in the context of this one.  Thanks a lot

djiatsaf-st commented:
> Hello @djiatsaf-st , could you please be so kind to add a flash partioning for WBA6 (nucleo and DK) ?
> 
> I could make a dedicated PR, but it will be better making these changes in the context of this one. Thanks a lot

Hello @asm5878, 

The PR has been updated with the changes for WBA6(nucleo and DK)  added.
